crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,321 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::bigtable::proto::bigtable::v2::bigtable_client::BigtableClient as BigtableInternalClient;
+use crate::bigtable::proto::bigtable::v2::mutate_rows_request::Entry;
+use crate::bigtable::proto::bigtable::v2::mutation::SetCell;
+use crate::bigtable::proto::bigtable::v2::read_rows_response::cell_chunk::RowStatus;
+use crate::bigtable::proto::bigtable::v2::{
+    mutation, MutateRowsRequest, MutateRowsResponse, Mutation, ReadRowsRequest, RowSet,
+};
+use crate::{Checkpoint, KeyValueStore, TransactionData};
+use anyhow::Result;
+use async_trait::async_trait;
+use gcp_auth::{Token, TokenProvider};
+use http::{HeaderValue, Request, Response};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+use std::task::{Context, Poll};
+use std::time::Duration;
+use sui_types::base_types::{SequenceNumber, TransactionDigest};
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::object::Object;
+use sui_types::storage::ObjectKey;
+use tonic::body::BoxBody;
+use tonic::codegen::Service;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Streaming;
+
+const COLUMN_FAMILY_NAME: &str = "sui";
+const COLUMN_QUALIFIER: &str = "";
+const OBJECTS_TABLE: &str = "objects";
+const CHECKPOINTS_TABLE: &str = "checkpoints";
+const TRANSACTIONS_TABLE: &str = "transactions";
+
+#[derive(Clone)]
+struct AuthChannel {
+    channel: Channel,
+    policy: String,
+    token_provider: Option<Arc<dyn TokenProvider>>,
+    token: Arc<RwLock<Option<Arc<Token>>>>,
+}
+
+#[derive(Clone)]
+pub struct BigTableClient {
+    table_prefix: String,
+    client: BigtableInternalClient<AuthChannel>,
+}
+
+#[async_trait]
+impl KeyValueStore for BigTableClient {
+    async fn get_objects(&mut self, object_keys: &[ObjectKey]) -> Result<Vec<Object>> {
+        let keys: Result<_, _> = object_keys.iter().map(|key| bcs::to_bytes(&key)).collect();
+        self.multi_get(OBJECTS_TABLE, keys?).await
+    }
+
+    async fn get_transactions(
+        &mut self,
+        transactions: &[TransactionDigest],
+    ) -> Result<Vec<TransactionData>> {
+        let keys = transactions.iter().map(|tx| tx.inner().to_vec()).collect();
+        self.multi_get(TRANSACTIONS_TABLE, keys).await
+    }
+
+    async fn get_checkpoint(
+        &mut self,
+        sequence_number: SequenceNumber,
+    ) -> Result<Option<Checkpoint>> {
+        let key = sequence_number.value().to_be_bytes().to_vec();
+        let mut result = self.multi_get(CHECKPOINTS_TABLE, vec![key]).await?;
+        Ok(result.pop())
+    }
+
+    async fn save_objects(&mut self, objects: &[&Object]) -> Result<()> {
+        let mut items = Vec::with_capacity(objects.len());
+        for object in objects {
+            let object_key = ObjectKey(object.id(), object.version());
+            items.push((bcs::to_bytes(&object_key)?, object));
```

Comment:
We can't use bcs for the object keys since bcs serializing integers in little endian it would throw off the lexicographical sorting that bigtable does. Instead lets manually construct the key ourself, skipping any serialization step as:

[32 byte object id] + [8 byte u64 version as big endian] for a fixed key size of 40 bytes

Reply:
good catch!